master/internal/experiment.go

```diff
@@ -490,6 +429,72 @@ func (e *experiment) Receive(ctx *actor.Context) error {
 	return nil
 }
 
+// snapshot snapshots the experiment and saves it in the database.
+func (e *experiment) snapshot(t trialSnapshot) error {
+	if ss, err := e.Snapshot(); err != nil {
+		return errors.Wrap(err, "failed to snapshot experiment")
+	} else if err := e.db.SaveSnapshot(e.ID, t.trialID, t.requestID, ss, t.snapshot); err != nil {
+		return errors.Wrap(err, "failed to persist experiment snapshot")
+	}
+	return nil
+}
+
+// restore restores this experiment from a snapshot in the database if it exists.
+func (e *experiment) restore() error {
+	switch b, err := e.db.ExperimentSnapshot(e.ID); {
+	case errors.Cause(err) == db.ErrNotFound:
+		log.WithField("experiment-id", e.ID).Info("no snapshot found")
+	case err != nil:
+		return errors.Wrap(err, "failed to retrieve experiment snapshot")
+	default:
+		if err := e.Restore(b); err != nil {
+			return errors.Wrap(err, "failed to restore experiment snapshot")
+		}
+	}
+	return nil
+}
+
+func (e *experiment) restoreTrial(
+	ctx *actor.Context, l *log.Entry, t *trial,
+) (terminal bool, err error) {
+	switch trial, err := t.db.TrialByExperimentAndRequestID(t.experiment.ID, t.create.RequestID); {
+	case errors.Cause(err) == db.ErrNotFound:
+		l.Info("trial was never previously allocated")
+		return false, nil
+	case err != nil:
+		return false, errors.Wrap(err, "failed to retrieve trial")
+	default:
+		if _, ok := model.TerminalStates[trial.State]; ok {
+			return true, nil
+		}
+		t.processID(trial.ID)
+	}
+
+	switch snapshot, err := t.db.TrialSnapshot(t.experiment.ID, t.create.RequestID); {
+	case errors.Cause(err) == db.ErrNotFound:
+		// Since the experiment saves the first snapshot upon receiving the trialCreated message
+		// from a trial, this can only happen if the master dies between when the trial saves itself
+		// to the database and the trialCreated message is received and handled. If we're here, the
+		// easiest fix is to just replay the trial created message.
+		l.Info("trial was previously allocated but had no snapshot")
+		if err = t.tellWithSnapshot(ctx, ctx.Self(), func(s trialSnapshot) interface{} {
+			return trialCreated{create: t.create, trialID: t.id, trialSnapshot: s}
+		}); err != nil {
+			return false, errors.Wrap(err, "failed to snapshot trial after missing")
+		}
+		return false, nil
+	case err != nil:
+		return false, errors.Wrap(err, "failed to retrieve trial snapshot")
+	default:
+		if err := t.Restore(snapshot); err != nil {
+			return false, errors.Wrap(err, "failed to restore trial snapshot")
+		}
+		l.Infof("loading trial at %d total batches processed", t.sequencer.TotalBatchesProcessed)
+	}
+
+	return false, errors.Wrap(t.reset(), "failed to reset trial in restore")
+}
+
 func (e *experiment) processOperations(
```

Comment:
A few thoughts about this function.   It kind of feels like we are hesitating from getting rid of the last remnants of the event log with this pattern of emitting searcher ops and triggering certain behaviors based on which events come out.

1. Can we separate the searcher operations which are "to the trial" from the searcher operations that are "to the experiment"?  Without the event log, it doesn't really make sense to have the searcher mix trial creation messages with train and validate messages.  The searcher could just call a function which creates a new trial the instant it decides to it wants one (no trialCreated message or trialCreated replays)
2. If we do step 1, the only searcher operations that remain are messages that build up the state of the workload sequencer (in other words, all operations should be tied to a RequestID), and then the experiment would literally not care about those *at all*.  It would just relay them to whatever RequestID was associated with the operations.  I think you would not need this processOperations() function at all, or `e.restored`.

Reply:
can we do this in a follow up? this pr is already massive and this is a good separate chunk of work to me. i'm totally ok with doing it, just my brain is about max out on concurrent large changes.

Reply:
I'm slightly concerned that the risk of breaking customers is much greater if we don't fully get rid of the old system.  Being in a halfway state seems to me to be harder to verify that things are correct as written, and unless we land both changes in a single release cycle we run two separate risks of breaking users.  I don't feel strongly about that though.  If you'd like to split it up we can do that.

But the one case where I would say we can't split it up is: if the follow up work would modify the nature of the snapshots, let's do it now so we don't need a snapshot migration step in between.  I'm afraid this actually *is* this case since the searcher operations would be changing, and that's one of the things you are saving in the snapshot.